error "cannot find -lopenblas: No such file or directory" on python 3.12

Hello,

I get this error: "/home/runner/micromamba/envs/test-env/compiler_compat/ld: cannot find -lopenblas: No such file or directory" when trying to build pylandstats. It seems that this only happens in Python 3.12 - see the logs [here](
https://github.com/martibosch/pylandstats/actions/runs/8899499863/job/24439006700). Any clues on how to solve this?

Thank you. Best,
Martí

--- a/pythran/config.py
+++ b/pythran/config.py
@@ -160,10 +160,8 @@ def make_extension(python, **extra):
     cfg = init_cfg('pythran.cfg',
                    'pythran-{}.cfg'.format(sys.platform),
                    '.pythranrc',
-                   extra.get('config', None))
+                   extra.pop('config', None))
 
-    if 'config' in extra:
-        extra.pop('config')
 
     def parse_define(define):
         index = define.find('=')

--- a/pythran/dist.py
+++ b/pythran/dist.py
@@ -4,6 +4,8 @@ This modules contains a distutils extension mechanism for Pythran
 '''
 
 import pythran.config as cfg
+from pythran.tables import blas_requires
+from pythran.utils import cxxid
 
 from collections import defaultdict
 try:
@@ -12,6 +14,7 @@ except ImportError:
     from collections import Iterable
 import os.path
 import os
+import re
 
 try:
     from distutils.command.build_ext import build_ext as LegacyBuildExt
@@ -140,23 +143,50 @@ class PythranBuildExt(PythranBuildExtMixIn, LegacyBuildExt, metaclass=PythranBui
     pass
 
 
+blas_requirements = {"/".join(map(cxxid, elem)) for elem in blas_requires}
+
+includes_matcher = re.compile(r'^#include <pythonic/include/(.*)\.hpp>$',
+                              re.MULTILINE)
+
+def requires_blas(source):
+    if not os.path.exists(source) or os.path.splitext(source)[1] != ".cpp":
+        return False  # conservative
+
+    with open(source) as fd:
+        content = fd.read()
+
+    return not blas_requirements.isdisjoint(includes_matcher.findall(content))
+
+
 class PythranExtension(Extension):
     '''
     Description of a Pythran extension
 
-    Similar to distutils.core.Extension except that the sources are .py files
+    Similar to setuptools.extension.Extension except that the sources are .py files
     They must be processable by pythran, of course.
 
     The compilation process ends up in a native Python module.
     '''
 
     def __init__(self, name, sources, *args, **kwargs):
+        self._kwargs = kwargs.copy()
+
+        if all(not requires_blas(source) for source in sources):
+            # Inserting at head so that user-specified config in CLI takes
+            # precedence.
+            kwargs['config'] = ['compiler.blas=none'] + kwargs.get('config', [])
+
         cfg_ext = cfg.make_extension(python=True, **kwargs)
         self.cxx = cfg_ext.pop('cxx', None)
         self.cc = cfg_ext.pop('cc', None)
-        self._sources = sources
         Extension.__init__(self, name, sources, *args, **cfg_ext)
         self.__dict__.pop("sources", None)
+
+    def _update_blas_requirements(self, source):
+        if requires_blas(source):
+            cfg_ext = cfg.make_extension(python=True, **self._kwargs)
+            for k, v in cfg_ext.items():
+                setattr(self, k, v)
 
     @property
     def sources(self):
@@ -178,9 +208,13 @@ class PythranExtension(Extension):
                     module_name = self.name
                 tc.compile_pythranfile(source, output_file,
                                        module_name, cpponly=True)
+
+            self._update_blas_requirements(output_file)
             cxx_sources.append(output_file)
         return cxx_sources
 
     @sources.setter
     def sources(self, sources):
         self._sources = sources
+        for source in sources:
+            self._update_blas_requirements(source)

--- a/pythran/intrinsic.py
+++ b/pythran/intrinsic.py
@@ -67,6 +67,7 @@ class Intrinsic(object):
                                        lambda call: UNKNOWN_RANGE)
         self.return_range_content = kwargs.get("return_range_content",
                                                lambda c: UNKNOWN_RANGE)
+        self.requires_blas = kwargs.get("requires_blas", False)
 
     def isliteral(self):
         return False

--- a/pythran/pythonic/include/operator_/imatmul.hpp
+++ b/pythran/pythonic/include/operator_/imatmul.hpp
@@ -1,8 +1,8 @@
 #ifndef PYTHONIC_INCLUDE_OPERATOR_IMATMUL_HPP
 #define PYTHONIC_INCLUDE_OPERATOR_IMATMUL_HPP
 
-#include "pythonic/include/utils/functor.hpp"
 #include "pythonic/include/numpy/dot.hpp"
+#include "pythonic/include/utils/functor.hpp"
 
 PYTHONIC_NS_BEGIN
 
@@ -12,10 +12,10 @@ namespace operator_
   template <class A, class B>
   A imatmul(A const &a, B &&b);
   template <class A, class B>
-  A imatmul(A &a, B &&b);
+  A &imatmul(A &a, B &&b);
 
   DEFINE_FUNCTOR(pythonic::operator_, imatmul);
-}
+} // namespace operator_
 PYTHONIC_NS_END
 
 #endif

--- a/pythran/pythonic/numpy/vdot.hpp
+++ b/pythran/pythonic/numpy/vdot.hpp
@@ -3,11 +3,11 @@
 
 #include "pythonic/include/numpy/vdot.hpp"
 
-#include "pythonic/utils/functor.hpp"
-#include "pythonic/types/ndarray.hpp"
-#include "pythonic/numpy/dot.hpp"
 #include "pythonic/numpy/asarray.hpp"
 #include "pythonic/numpy/conjugate.hpp"
+#include "pythonic/numpy/dot.hpp"
+#include "pythonic/types/ndarray.hpp"
+#include "pythonic/utils/functor.hpp"
 
 PYTHONIC_NS_BEGIN
 
@@ -21,16 +21,14 @@ namespace numpy
   {
     if (types::is_complex<typename U::dtype>::value &&
         types::is_complex<typename V::dtype>::value) {
-      puts("complex");
       return functor::dot{}(functor::asarray{}(functor::conjugate{}(u)).flat(),
                             functor::asarray{}(v).flat());
     } else {
-      puts("simplex");
       return functor::dot{}(functor::asarray{}(u).flat(),
                             functor::asarray{}(v).flat());
     }
   }
-}
+} // namespace numpy
 PYTHONIC_NS_END
 
 #endif

--- a/pythran/pythonic/operator_/imatmul.hpp
+++ b/pythran/pythonic/operator_/imatmul.hpp
@@ -3,8 +3,8 @@
 
 #include "pythonic/include/operator_/imatmul.hpp"
 
-#include "pythonic/utils/functor.hpp"
 #include "pythonic/numpy/dot.hpp"
+#include "pythonic/utils/functor.hpp"
 
 PYTHONIC_NS_BEGIN
 
@@ -19,10 +19,10 @@ namespace operator_
   template <class A, class B>
   A &imatmul(A &a, B &&b)
   {
-    return a = numpy::functor::dot(a,
-                                   std::forward<B>(b)); // FIXME: improve that
+    return a = numpy::functor::dot{}(a,
+                                     std::forward<B>(b)); // FIXME: improve that
   }
-}
+} // namespace operator_
 PYTHONIC_NS_END
 
 #endif

--- a/pythran/tables.py
+++ b/pythran/tables.py
@@ -3586,8 +3586,8 @@ MODULES = {
         "complex256": ConstFunctionIntr(signature=_complex_signature),
         "conj": ConstMethodIntr(signature=_numpy_unary_op_signature),
         "conjugate": ConstMethodIntr(signature=_numpy_unary_op_signature),
-        "convolve": ConstMethodIntr(),
-        "correlate": ConstMethodIntr(),
+        "convolve": ConstMethodIntr(requires_blas=True),
+        "correlate": ConstMethodIntr(requires_blas=True),
         "copy": ConstMethodIntr(signature=_numpy_array_signature),
         "copyto": FunctionIntr(
             argument_effects=[UpdateEffect(), ReadEffect(),
@@ -3709,7 +3709,7 @@ MODULES = {
         "diff": ConstFunctionIntr(),
         "digitize": ConstFunctionIntr(),
         "divide": UFunc(BINARY_UFUNC),
-        "dot": ConstMethodIntr(),
+        "dot": ConstMethodIntr(requires_blas=True),
         "double": ConstFunctionIntr(signature=_float_signature),
         "dtype": ClassWithConstConstructor(CLASSES["dtype"]),
         "e": ConstantIntr(),
@@ -3779,7 +3779,7 @@ MODULES = {
         "indices": ConstFunctionIntr(),
         "inf": ConstantIntr(),
         "Inf": ConstantIntr(),
-        "inner": ConstFunctionIntr(),
+        "inner": ConstFunctionIntr(requires_blas=True),
         "insert": ConstFunctionIntr(),
         "interp": ConstFunctionIntr(),
         "intersect1d": ConstFunctionIntr(),
@@ -3817,7 +3817,7 @@ MODULES = {
         "lexsort": ConstFunctionIntr(),
         "linalg": {
             "norm": FunctionIntr(),
-            "matrix_power": ConstFunctionIntr(),
+            "matrix_power": ConstFunctionIntr(requires_blas=True),
         },
         "linspace": ConstFunctionIntr(),
         "log": ConstFunctionIntr(),
@@ -4071,7 +4071,7 @@ MODULES = {
         "ushort": ConstFunctionIntr(signature=_int_signature),
         "var": ConstMethodIntr(),
         "vectorize": ConstFunctionIntr(),
-        "vdot": ConstMethodIntr(),
+        "vdot": ConstMethodIntr(requires_blas=True),
         "vstack": ConstFunctionIntr(),
         "where": ConstFunctionIntr(),
         "zeros": ConstFunctionIntr(args=('shape', 'dtype'),
@@ -4408,8 +4408,12 @@ MODULES = {
         "__lshift__": ConstFunctionIntr(
             signature=_numpy_int_binary_op_signature
         ),
-        "matmul": ConstFunctionIntr(signature=_operator_mul_signature),
-        "__matmul__": ConstFunctionIntr(signature=_operator_mul_signature),
+        "matmul": ConstFunctionIntr(signature=_operator_mul_signature,
+                                    requires_blas=True),
+        "__matmul__": ConstFunctionIntr(signature=_operator_mul_signature,
+                                        requires_blas=True),
+        "imatmul": MethodIntr(update_effects, requires_blas=True),
+        "__imatmul__": MethodIntr(update_effects, requires_blas=True),
         "mod": ConstFunctionIntr(signature=_operator_mod_signature),
         "__mod__": ConstFunctionIntr(signature=_operator_mod_signature),
         "mul": ConstFunctionIntr(signature=_operator_mul_signature),
@@ -4737,6 +4741,21 @@ def save_attribute(elements, module_path):
 
 
 save_attribute(MODULES, ())
+
+blas_requires = set()
+
+def save_blas_requires(elements, module_path):
+    """ Recursively save attributes with module name and signature. """
+    for elem, signature in elements.items():
+        if isinstance(signature, dict):  # Submodule case
+            save_blas_requires(signature, module_path + (elem,))
+        elif signature.requires_blas:
+            blas_requires.add(module_path + (elem,))
+        elif isinstance(signature, Class):
+            save_blas_requires(signature.fields, module_path + (elem,))
+
+save_blas_requires(MODULES, ())
+
 
 # patch beniget with pythran-specific builtins
 import beniget

--- a/pythran/tests/test_advanced.py
+++ b/pythran/tests/test_advanced.py
@@ -287,6 +287,14 @@ def combiner_on_empty_list():
             numpy.array([[0., 2.], [1., 3.]]),
             matmul_operator=[NDArray[float, :,:], NDArray[float, :,:]])
 
+    def test_imatmul_operator(self):
+        code = 'def imatmul_operator(x, y): x @= y; return x'
+        self.run_test(
+            code,
+            numpy.array([[1., 1.], [2., 2.]]),
+            numpy.array([[0., 2.], [1., 3.]]),
+            imatmul_operator=[NDArray[float, :,:], NDArray[float, :,:]])
+
     def test_generator_handler_name(self):
         code = '''
             def foo(x):

--- a/pythran/tests/test_advanced.py
+++ b/pythran/tests/test_advanced.py
@@ -5,6 +5,7 @@ from unittest import skip, skipIf
 import numpy
 
 from pythran.typing import *
+from packaging.version import Version
 
 class TestAdvanced(TestEnv):
 
@@ -287,6 +288,7 @@ def combiner_on_empty_list():
             numpy.array([[0., 2.], [1., 3.]]),
             matmul_operator=[NDArray[float, :,:], NDArray[float, :,:]])
 
+    @skipIf(Version(numpy.__version__) <= Version('1.26'), "Not supported upstream")
     def test_imatmul_operator(self):
         code = 'def imatmul_operator(x, y): x @= y; return x'
         self.run_test(

--- a/pythran/tests/test_distutils/b.py
+++ b/pythran/tests/test_distutils/b.py
@@ -1,0 +1,4 @@
+#pythran export b(int)
+import numpy
+def b(n): return numpy.ones((n, n)) @ numpy.ones((n, n))
+

--- a/pythran/tests/test_distutils/setup.py
+++ b/pythran/tests/test_distutils/setup.py
@@ -2,9 +2,10 @@ from distutils.core import setup
 from pythran.dist import PythranExtension, PythranBuildExt
 
 module1 = PythranExtension('demo', sources = ['a.py'])
+module2 = PythranExtension('demo', sources = ['b.py'])
 
 setup(name = 'demo',
       version = '1.0',
       description = 'This is a demo package',
       cmdclass={"build_ext": PythranBuildExt},
-      ext_modules = [module1])
+      ext_modules = [module1, module2])

--- a/pythran/tests/test_ndarray.py
+++ b/pythran/tests/test_ndarray.py
@@ -2,8 +2,6 @@ from pythran.tests import TestEnv
 from pythran.typing import NDArray, Tuple, List
 
 import numpy
-from distutils.version import LooseVersion
-
 import unittest
 
 try:

--- a/pythran/tests/test_numpy_fft.py
+++ b/pythran/tests/test_numpy_fft.py
@@ -2,7 +2,7 @@ import unittest
 from pythran.tests import TestEnv
 import numpy
 from pythran.typing import NDArray
-from distutils.version import LooseVersion
+from packaging.version import Version
 
 
 class TestNumpyRFFT(TestEnv):
@@ -18,7 +18,7 @@ class TestNumpyRFFT(TestEnv):
     def test_rfft_3(self):
         self.run_test("def test_rfft_3(x,n): from numpy.fft import rfft ; return rfft(x,n)", numpy.arange(0,8.),7, test_rfft_3=[NDArray[float,:],int])
 
-    @unittest.skipIf(LooseVersion(numpy.__version__) >= '2', "see https://github.com/numpy/numpy/issues/26349")
+    @unittest.skipIf(Version(numpy.__version__) >= Version('2'), "see https://github.com/numpy/numpy/issues/26349")
     def test_rfft_4(self):
         self.run_test("def test_rfft_4(x,n): from numpy.fft import rfft ; return rfft(x,n)", numpy.arange(0,8.),6, test_rfft_4=[NDArray[float,:],int])
 
@@ -132,7 +132,7 @@ class TestNumpyFFTN(TestEnv):
 
     # Various norms
 
-    @unittest.skipIf(LooseVersion(numpy.__version__) <'1.20', "introduced in 1.20")
+    @unittest.skipIf(Version(numpy.__version__) < Version('1.20'), "introduced in 1.20")
     def test_fftn_12(self):
         self.run_test("def test_fftn_12(x): from numpy.fft import fftn ; return fftn(x, (6,), norm='backward')",
                       numpy.arange(0,8), test_fftn_12=[NDArray[int,:]])
@@ -140,7 +140,7 @@ class TestNumpyFFTN(TestEnv):
         self.run_test("def test_fftn_13(x): from numpy.fft import fftn ; return fftn(x, (8,), norm='ortho')",
                       numpy.arange(0,8.), test_fftn_13=[NDArray[float,:]])
 
-    @unittest.skipIf(LooseVersion(numpy.__version__) <'1.20', "introduced in 1.20")
+    @unittest.skipIf(Version(numpy.__version__) < Version('1.20'), "introduced in 1.20")
     def test_fftn_14(self):
         self.run_test("def test_fftn_14(x): from numpy.fft import fftn ; return fftn(x, (10,), norm='forward')",
                       numpy.arange(0, 8.) + 1.j, test_fftn_14=[NDArray[complex,:]])
@@ -258,7 +258,7 @@ class TestNumpyIHFFT(TestEnv):
     def test_ihfft_3(self):
         self.run_test("def test_ihfft_3(x,n): from numpy.fft import ihfft ; return ihfft(x,n)", numpy.arange(0,8.),7, test_ihfft_3=[NDArray[float,:],int])
 
-    @unittest.skipIf(LooseVersion(numpy.__version__) >= '2', "see https://github.com/numpy/numpy/issues/26349")
+    @unittest.skipIf(Version(numpy.__version__) >= Version('2'), "see https://github.com/numpy/numpy/issues/26349")
     def test_ihfft_4(self):
         self.run_test("def test_ihfft_4(x,n): from numpy.fft import ihfft ; return ihfft(x,n)", numpy.arange(0,8.),6, test_ihfft_4=[NDArray[float,:],int])
 

